src/wallet/rpcdump.cpp

```diff
@@ -854,211 +941,167 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
-                throw JSONRPCError(RPC_WALLET_ERROR, "Error adding p2sh redeemScript to wallet");
-            }
-
-            // Now set script to the redeemScript so we parse the inner script as P2WSH or P2WPKH below
-            script = redeemScript;
-            ExtractDestination(script, dest);
+            auto parsed_redeemscript = ParseHex(strRedeemScript);
+            import_data.redeemscript = MakeUnique<CScript>(parsed_redeemscript.begin(), parsed_redeemscript.end());
         }
-
-        // (P2SH-)P2WSH
-        if (!witness_script_hex.empty() && script.IsPayToWitnessScriptHash()) {
+        if (witness_script_hex.size()) {
             if (!IsHex(witness_script_hex)) {
                 throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Invalid witness script: must be hex string");
             }
-
-            // Generate the scripts
-            std::vector<unsigned char> witness_script_parsed(ParseHex(witness_script_hex));
-            CScript witness_script = CScript(witness_script_parsed.begin(), witness_script_parsed.end());
-            CScriptID witness_id(witness_script);
-
-            // Check that the witnessScript and scriptPubKey match
-            if (GetScriptForDestination(WitnessV0ScriptHash(witness_script)) != script) {
-                throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "The witnessScript does not match the scriptPubKey or redeemScript");
-            }
-
-            // Add the witness script as watch only only if it is not for P2SH-P2WSH
-            if (!scriptpubkey_script.IsPayToScriptHash() && !pwallet->AddWatchOnly(witness_script, timestamp)) {
-                throw JSONRPCError(RPC_WALLET_ERROR, "Error adding address to wallet");
-            }
-
-            if (!pwallet->HaveCScript(witness_id) && !pwallet->AddCScript(witness_script)) {
-                throw JSONRPCError(RPC_WALLET_ERROR, "Error adding p2wsh witnessScript to wallet");
-            }
-
-            // Now set script to the witnessScript so we parse the inner script as P2PK or P2PKH below
-            script = witness_script;
-            ExtractDestination(script, dest);
-            allow_p2wpkh = false; // P2WPKH cannot be embedded in P2WSH
+            auto parsed_witnessscript = ParseHex(witness_script_hex);
+            import_data.witnessscript = MakeUnique<CScript>(parsed_witnessscript.begin(), parsed_witnessscript.end());
         }
-
-        // (P2SH-)P2PK/P2PKH/P2WPKH
-        if (dest.type() == typeid(CKeyID) || dest.type() == typeid(WitnessV0KeyHash)) {
-            if (!allow_p2wpkh && dest.type() == typeid(WitnessV0KeyHash)) {
-                throw JSONRPCError(RPC_INVALID_PARAMETER, "P2WPKH cannot be embedded in P2WSH");
+        for (size_t i = 0; i < pubKeys.size(); ++i) {
+            const auto& str = pubKeys[i].get_str();
+            if (!IsHex(str)) {
+                throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Pubkey must be a hex string");
```

Comment:
Done. I thought about doing this before but worried it would complicate writing tests. Since you've volunteered to help out with that, I don't assume that'll be a problem :p